### PR TITLE
RevEng: Add HasIndex() Fluent API

### PR DESCRIPTION
**Pull request by lajones**
Fix for #2840.

Also tidied up some of the Configuration classes.

### CODE REVIEW

File: src/EntityFramework.Relational.Design/Internal/Configuration/IndexConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        public virtual bool HasAttributeEquivalent { get; }
+        public virtual string For { get; }
+
+        public virtual string FluentApi
+        {
+            get
+            {
+                var sb = new StringBuilder();
+                sb.Append("HasIndex(");
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/Internal/Configuration/IndexConfiguration.cs]: Nit: refactoring safeguard: use `nameof()`

File: test/EntityFramework.MicrosoftSqlServer.Design.FunctionalTests/ReverseEngineering/ExpectedResults/E2E_AllFluentApi/SqlServerReverseEngineerTestE2EContext.expected [hunk 2]
```diff
@@ -102,6 +102,8 @@ namespace E2ETest.Namespace
             {
                 entity.HasKey(e => new { e.OneToOneFKToUniqueKeyDependentID1, e.OneToOneFKToUniqueKeyDependentID2 });
 
+                entity.HasIndex(e => new { e.OneToOneFKToUniqueKeyDependentFK1, e.OneToOneFKToUniqueKeyDependentFK2 }).HasName("UK_OneToOneFKToUniqueKeyDependent").IsUnique();
```
Comment (natemcmaster) [inline test/EntityFramework.MicrosoftSqlServer.Design.FunctionalTests/ReverseEngineering/ExpectedResults/E2E_AllFluentApi/SqlServerReverseEngineerTestE2EContext.expected]: Nit: line breaks for each call in the chain?

Comment (lajones) [inline test/EntityFramework.MicrosoftSqlServer.Design.FunctionalTests/ReverseEngineering/ExpectedResults/E2E_AllFluentApi/SqlServerReverseEngineerTestE2EContext.expected]: I'll add this to #3394.

File: src/EntityFramework.Relational.Design/Internal/Configuration/IndexConfiguration.cs (hunk 1 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/Internal/Configuration/IndexConfiguration.cs]: Good point. Will do.

**Comment by natemcmaster** [comment]
:shipit: (when checkins allowed again)

**Comment by lajones** [comment]
Checked into release (ec30dd5) and dev (aa2ee57).

**Pull request closed without merging**